### PR TITLE
PWN-7341 - Add AlchemyApi, provide methods for work with tokens

### PR DESCRIPTION
**Pull request by Davran331122**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-7341

### CODE REVIEW

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/AlchemyService.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import retrofit2.http.Body
+import retrofit2.http.Headers
+import retrofit2.http.POST
+import retrofit2.http.Url
+import java.net.URI
+
+interface AlchemyService {
+    @POST
+    @Headers("Content-Type: application/json","Accept: application/json")
+    suspend fun launch(@Url uri: URI, @Body jsonRpc: String): RpcResponse
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/AlchemyService.kt:13]: retrofit can work with URI?

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/NetworkModule.kt [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import org.p2p.ethereumkit.external.core.EthereumNetworkEnvironment
+import org.p2p.ethereumkit.external.core.GsonProvider
+import retrofit2.Retrofit
+import retrofit2.converter.gson.GsonConverterFactory
+import retrofit2.converter.scalars.ScalarsConverterFactory
+import java.util.logging.Logger
+
+private const val ETH_PREFIX = "ethereum"
+
+object NetworkModule {
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/NetworkModule.kt:20]: EtheriumNetworkModule

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/response/TokenBalancesResponse.kt [hunk 3]
```diff
@@ -0,0 +1,30 @@
+package org.p2p.ethereumkit.external.api.response
+
+import com.google.gson.annotations.SerializedName
+import org.p2p.ethereumkit.internal.models.EthAddress
+import java.math.BigInteger
+
+internal data class TokenBalancesResponse(
+    @SerializedName("address")
+    val address: EthAddress? = null,
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/response/TokenBalancesResponse.kt:9]: EthAddressResponse

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/response/TokenBalancesResponse.kt [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    val contractAddress: EthAddress,
+    @SerializedName("tokenBalance")
+    val tokenBalance: BigInteger
+)
+
+internal data class TokenMetadataResponse(
+    @SerializedName("decimals")
+    val decimals: Int,
+    @SerializedName("logo")
+    val logo: String? = null,
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/response/TokenBalancesResponse.kt:25]: logoUrl?

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/response/TokenBalancesResponse.kt [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    @SerializedName("name")
+    val name: String? = null,
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/response/TokenBalancesResponse.kt:27]: tokenName

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/response/TokenBalancesResponse.kt [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    @SerializedName("symbol")
+    val symbol: String? = null
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/response/TokenBalancesResponse.kt:29]: tokenSymbol

File: ethereumkit/src/test/java/org/p2p/ethereumkit/TokenBalancesTest.kt [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+            val tokenLogo = tokenMetadata.logo
+            val tokenBalance = balance.balances.first { it.contractAddress == address }.tokenBalance
+            stringBuilder.append("Token Name: $tokenName ")
+            stringBuilder.append("Token Symbol: $tokenSymbol ")
+            stringBuilder.append("Token Balance: $tokenBalance ")
+            stringBuilder.append("Token Decimals: $tokenDecimals ")
+            stringBuilder.append("Token logo: $tokenLogo ")
+            stringBuilder.append("\n")
+        }
+        println(stringBuilder.toString())
```
Comment (gslevinkov) [inline ethereumkit/src/test/java/org/p2p/ethereumkit/TokenBalancesTest.kt:56]: don't forget to remove println everywhere

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/model/EthTokenKeyProvider.kt [hunk 8]
```diff
@@ -1,9 +1,18 @@
...
-    val privateKey: BigInteger
-)
+    private val seedPhrase: List<String>
+) {
+
+    private val mnemonic = Mnemonic().toSeed(seedPhrase)
+
+    val address: EthAddress = Signer.Companion.address(mnemonic, Chain.Ethereum)
+
+    val privateKey: BigInteger = Signer.privateKey(mnemonic,Chain.Ethereum)
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/model/EthTokenKeyProvider.kt:17]: why privateKey is BigInteger? 0-0

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/core/GsonProvider.kt [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+import org.p2p.ethereumkit.internal.network.OptionalTypeAdapter
+import java.math.BigInteger
+import java.util.Optional
+
+class GsonProvider {
+
+    var gson: Gson? = null
+
+    fun provide(): Gson {
+        return gson ?: GsonBuilder().also {
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/core/GsonProvider.kt:26]: apply would be good

also you can extract this also to a method and everything will look like this

fun provide(): Gson = gson ?: buildGson()

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/model/EthTokenKeyProvider.kt [hunk 10]
```diff
@@ -1,9 +1,18 @@
...
 import java.math.BigInteger
 
 class EthTokenKeyProvider(
-    val publicKey: EthAddress,
-    val privateKey: BigInteger
-)
+    private val seedPhrase: List<String>
+) {
+
+    private val mnemonic = Mnemonic().toSeed(seedPhrase)
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/model/EthTokenKeyProvider.kt:13]: explicit return type please

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/NetworkModule.kt [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+import org.koin.core.qualifier.named
+import org.koin.dsl.module
+import org.p2p.ethereumkit.external.core.EthereumNetworkEnvironment
+import org.p2p.ethereumkit.external.core.GsonProvider
+import retrofit2.Retrofit
+import retrofit2.converter.gson.GsonConverterFactory
+import retrofit2.converter.scalars.ScalarsConverterFactory
+import java.util.logging.Logger
+
+private const val ETH_PREFIX = "ethereum"
```
Comment (gslevinkov) [inline ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/NetworkModule.kt:18]: better to use enum, QUALIFIER_ETH_HTTP_CLIENT

**Review by gslevinkov — APPROVED** [review]

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/AlchemyService.kt (hunk 1 above)
Comment (Davran331122) [reply ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/AlchemyService.kt:13]: yeah, it's more comfortable to use URI)

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/response/TokenBalancesResponse.kt (hunk 3 above)
Comment (Davran331122) [reply ethereumkit/src/main/java/org/p2p/ethereumkit/external/api/response/TokenBalancesResponse.kt:9]: no, it's not response, it's wrapped primitive class

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/model/EthTokenKeyProvider.kt (hunk 8 above)
Comment (Davran331122) [reply ethereumkit/src/main/java/org/p2p/ethereumkit/external/model/EthTokenKeyProvider.kt:17]: I don't know, but it's BigInterger from our kit

**Review by kabramovp2p — APPROVED** [review]

File: ethereumkit/src/main/java/org/p2p/ethereumkit/external/EthereumModule.kt [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+import org.p2p.ethereumkit.external.core.EthereumNetworkEnvironment
+import org.p2p.ethereumkit.external.model.EthTokenKeyProvider
+import org.p2p.ethereumkit.external.repository.EthereumKitRepository
+import org.p2p.ethereumkit.external.repository.EthereumRepository
+
+object EthereumModule {
+
+    fun create(): Module = module {
+        single {
+            val seedPhrase = "apart approve black comfort steel spin real renew tone primary key cherry".split(" ")
```
[the remark this answers is not in the document]
Comment (Davran331122) [reply ethereumkit/src/main/java/org/p2p/ethereumkit/external/EthereumModule.kt:16]: we will provide it from our SeedPhraseProvider

**Pull request merged**